### PR TITLE
work on integrating OIDC provider workflow as a built-in plugin

### PR DESCRIPTION
**Pull request by indirectlylit**
### Summary

Builds off of @jredrejo's work in https://github.com/learningequality/kolibri-un-women-online-plugin/pull/7/

### Reviewer guidance

* Attempts to use the standard sign in page for OIDC sign-in
* Can be activated by running `kolibri plugin kolibri.plugins.oidc_provider_plugin enable`
* Mounts `openid-configuration` at the domain root

### References

…

----

### Contributor Checklist

PR process:

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/assets/src/state/modules/core/actions.js [hunk 1]
```diff
@@ -243,8 +243,14 @@ export function kolibriLogin(store, sessionPayload) {
   Lockr.set(UPDATE_MODAL_DISMISSED, false);
   return SessionResource.saveModel({ data: sessionPayload })
     .then(() => {
-      // Redirect on login
-      redirectBrowser();
+      // OIDC redirect
+      if (sessionPayload.next) {
+        redirectBrowser(sessionPayload.next);
```
Comment (indirectlylit) [inline kolibri/core/assets/src/state/modules/core/actions.js:248]: This is where redirect can occur. Unsure if this is actually what's supposed to happen next in the flow?

File: kolibri/deployment/default/settings/dev.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 from .base import *  # noqa isort:skip @UnusedWildImport
 
-INSTALLED_APPS += ["rest_framework_swagger"]  # noqa
+INSTALLED_APPS = list(INSTALLED_APPS) + ["rest_framework_swagger"]  # noqa F405
```
Comment (indirectlylit) [inline kolibri/deployment/default/settings/dev.py:7]: `INSTALLED_APPS` was a tuple

File: kolibri/plugins/oidc_provider_plugin/settings.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+INSTALLED_APPS = ["oidc_provider"]
+OIDC_LOGIN_URL = "/user/#/signin"
```
Comment (indirectlylit) [inline kolibri/plugins/oidc_provider_plugin/settings.py]: Note, switched from `LOGIN_URL` to `OIDC_LOGIN_URL`

Will this work as intended though? I'm unsure of how things will be affected by our language prefixes, and also potentially by the client-side routing

File: kolibri/plugins/user/assets/src/views/SignInPage/index.vue [hunk 4]
```diff
@@ -405,11 +410,15 @@
...
         if (this.formIsValid) {
-          this.kolibriLogin({
+          const sessionPayload = {
             username: this.username,
             password: this.password,
             facility: this.facilityId,
-          }).catch();
+          };
+          if (global.oidcProviderEnabled) {
+            sessionPayload['next'] = this.$route.query.next;
```
Comment (indirectlylit) [inline kolibri/plugins/user/assets/src/views/SignInPage/index.vue]: grab the `next` query parameter and add it to the payload

**Review by indirectlylit — COMMENTED** [review]
@jredrejo @rtibbles left some comments inline

File: kolibri/plugins/oidc_provider_plugin/settings.py (hunk 3 above)
Comment (jredrejo) [reply kolibri/plugins/oidc_provider_plugin/settings.py]: nope, it has to be `LOGIN_URL` . The provider library requires this name. We can do some monkey patching if you want to add a prefix.

File: kolibri/deployment/default/settings/dev.py (hunk 2 above)
Comment (rtibbles) [reply kolibri/deployment/default/settings/dev.py:7]: Ah, right, the plugin settings modification coerces it to a tuple.

File: kolibri/plugins/base.py [hunk 5]
```diff
@@ -181,6 +186,31 @@ def api_url_module(self):
...
+        Django application url module.
+
+        Do this separately for endpoints that need to be attached at the root.
+
+        URLs are by default accessed through Django's reverse lookups like
+        this::
+
+            reverse('kolibri:url_name')
+
+        By default this will be discovered based on the root_view_urls
```
Comment (rtibbles) [inline kolibri/plugins/base.py:201]: I would like a very big warning here, just to scare people.

File: kolibri/plugins/oidc_provider_plugin/api_urls.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from django.conf.urls import include
+from django.conf.urls import url
+from rest_framework import routers
+
+router = routers.SimpleRouter()
```
Comment (rtibbles) [inline kolibri/plugins/oidc_provider_plugin/api_urls.py:5]: Looks like this is doing nothing, as we are not registering any endpoints against the router.

File: kolibri/plugins/oidc_provider_plugin/root_urls.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+from django.conf.urls import url
+from rest_framework import routers
+
+from .views import ProviderInfoView
+
+router = routers.SimpleRouter()
```
Comment (rtibbles) [inline kolibri/plugins/oidc_provider_plugin/root_urls.py:6]: Ditto.

File: kolibri/plugins/user/assets/src/views/SignInPage/index.vue [hunk 8]
```diff
@@ -166,6 +168,14 @@
   import LanguageSwitcherFooter from '../LanguageSwitcherFooter';
   import FacilityModal from './FacilityModal';
 
+  // https://davidwalsh.name/query-string-javascript
+  function getUrlParameter(name) {
+    name = name.replace(/[[]/, '[').replace(/[\]]/, '\\]');
+    var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
```
Comment (rtibbles) [inline kolibri/plugins/user/assets/src/views/SignInPage/index.vue:174]: Thought we had this somewhere already, but I think it was in Hashi and then I later removed it. Can also polyfill the URL constructor too for this.

File: kolibri/plugins/user/assets/src/views/SignInPage/index.vue [hunk 9]
```diff
@@ -292,6 +302,17 @@
         }
         return { backgroundColor: this.$themeColors.brand.primary.v_900 };
       },
+      oidcProviderFlow() {
+        return global.oidcProviderEnabled && this.nextParam;
+      },
+      nextParam() {
```
Comment (rtibbles) [inline kolibri/plugins/user/assets/src/views/SignInPage/index.vue:308]: Django ordinarily has support for this, but I had removed it in our implementation - I actually kind of prefer handling it in the frontend, but we could reinstate the backend functionality if that was preferred.

File: kolibri/plugins/utils/urls.py [hunk 10]
```diff
@@ -10,13 +10,16 @@ def get_urls():
...
+        root_url_module = plugin_instance.root_url_module()
         instance_patterns = []
         # Normalize slug
         slug = plugin_instance.url_slug().lstrip("^").rstrip("/") + "/"
         if url_module:
             instance_patterns += i18n_patterns(url_module.urlpatterns, prefix=slug)
         if api_url_module:
             instance_patterns.append(url(slug + "api/", include(api_url_module)))
+        if root_url_module:
+            instance_patterns.append(url("", include(root_url_module)))
```
Comment (rtibbles) [inline kolibri/plugins/utils/urls.py:22]: What I had suggested was to not mount these within the plugin namespace (so they are doubly 'root' - both in terms of their point of mounting, and the Django namespace in which they are mounted).

Would want to extract this specific part into a separate 'root_urls' function that gets separately called and mounted in the `deployment.default.urls` configuration. This should allow us to remove a lot of the monkey patching above.

File: kolibri/utils/conf.py [hunk 11]
```diff
@@ -77,6 +77,7 @@
         "kolibri.plugins.style_guide",
         "kolibri.plugins.document_epub_render",
         "kolibri.plugins.default_theme",
+        "kolibri.plugins.oidc_provider_plugin",
```
Comment (rtibbles) [inline kolibri/utils/conf.py:80]: Do we want this to be the default during dev?

Comment (indirectlylit) [reply kolibri/utils/conf.py:80]: no, this was supposed to be  removed

File: kolibri/plugins/base.py (hunk 5 above)
Comment (indirectlylit) [reply kolibri/plugins/base.py:201]: added warning

File: kolibri/plugins/user/assets/src/views/SignInPage/index.vue (hunk 8 above)
Comment (indirectlylit) [reply kolibri/plugins/user/assets/src/views/SignInPage/index.vue:174]: looks like [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#Browser_compatibility) is the new hotness

Comment (rtibbles) [reply kolibri/plugins/user/assets/src/views/SignInPage/index.vue:174]: Yeah, when you parse a url with the URL constructor, the it generates a URLSearchParams object for the query params.

**Pull request merged**